### PR TITLE
[MoE] Add overridable hooks for TokenChoiceTopKRouter

### PR DESCRIPTION
**Pull request by jinsooihm**
`TokenChoiceTopKRouter` is a module with a large `forward`. 
Some models may want to override how the score is computed or how the normalization is done. 
Extracting and exposing them as overridable functions `_compute_scores` and `_normalize_topk_scores` would make it easier. 

## Test Plan
- `pytest -q tests/unit_tests/cpu/test_moe.py tests/unit_tests/cpu/test_router_gate_linear.py`
    - 8 passed

- Targeted pre-commit checks for `moe.py` and `test_moe.py`
    - All passed

 - `pyrefly check torchtitan/models/common/moe.py`
    - 0 errors

- Compared upstream/main (773e16e75) against 5f7de4008 using scripts/loss_compare.py, --debug.seed=42, --debug.deterministic, and 10 training steps.
    - Qwen3 MoE: loss and grad_norm bitwise identical for all 10 steps
    - DeepSeek V3: loss and grad_norm bitwise identical for all 10 steps
    - GPT-OSS: loss and grad_norm bitwise identical for all 10 steps
    - DeepSeek V4: loss and grad_norm bitwise identical for all 10 steps

### CODE REVIEW

**Review by tianyu-l — CHANGES_REQUESTED** [review]
Without context, hard to evaluate this PR.

E.g. for deepseek v4, we added https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/common/moe.py#L290

**Comment by jinsooihm** [comment]
@tianyu-l Updated after offline discussion.

File: torchtitan/models/common/activation.py
Comment (tianyu-l) [inline torchtitan/models/common/activation.py]: maybe
```suggestion
class BinaryActivationFn(Function[torch.Tensor], ABC):
```

File: torchtitan/models/common/config_utils.py [hunk 1]
```diff
@@ -334,8 +335,9 @@ def make_router_config(
     num_experts: int,
     gate_param_init: dict[str, Callable],
     top_k: int = 1,
-    score_func: Literal["sigmoid", "softmax", "sqrtsoftplus"] = "sigmoid",
+    score_func: UnaryActivationFn.Config | None = None,
```
Comment (tianyu-l) [inline torchtitan/models/common/config_utils.py]: would it ever be None?

File: torchtitan/models/common/moe.py [hunk 2]
```diff
@@ -187,8 +191,9 @@ class Config(Module.Config):
         num_experts: int
         gate: RouterGateLinear.Config
         top_k: int = 1
-        score_func: Literal["softmax", "sigmoid", "sqrtsoftplus"] = "sigmoid"
+        score_func: UnaryActivationFn.Config = field(default_factory=Sigmoid.Config)
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: can it not have a default?

**Review by tianyu-l — APPROVED** [review]
please address remaining comments

File: torchtitan/models/common/config_utils.py (hunk 1 above)
Comment (jinsooihm) [reply torchtitan/models/common/config_utils.py]: All models specify the function, only the tests call it without specifying which defaults to sigmoid. We can remove it.
